File: 328-odd-even-linked-list/solution.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function oddEvenList(head: ListNode | null): ListNode | null {
  if (!head) return head;

  let now = head;
  let evenHead: ListNode | null = null;
  let evenNow: ListNode | null = null;
  while (true) {
    if (!evenHead) evenHead = now.next;
    if (evenNow !== null) evenNow.next = now.next;
    evenNow = now.next;
    now.next = now?.next?.next || null;
    if (evenNow !== null) evenNow.next = null;
    if (now.next === null) {
      now.next = evenHead;
      break;
    }
    now = now.next;
  }

  return head;
}

File: 328-odd-even-linked-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if(head == NULL) return NULL;
        ListNode* oddHead = NULL;
        ListNode* oddNow = NULL;
        ListNode* evenHead = NULL;
        ListNode* evenNow = NULL;
        ListNode* now = head;
        int idx = 1;
        while(now != NULL) {
            if(idx == 1) {
                oddHead = now; oddNow = now;
            } else if(idx == 2) {
                evenHead = now; evenNow = now;
            } else if(idx % 2 == 1) {
                oddNow->next = now; oddNow = now;
            } else {
                evenNow->next = now; evenNow = now;
            }
            now = now->next;
            idx++;
        }
        oddNow->next = evenHead;
        if(evenNow != NULL) evenNow->next = NULL; // IMPORTANT!!!
        return oddHead;
    }
};
